src/shared/serialize.c

```diff
@@ -116,6 +116,26 @@ int serialize_fd(FILE *f, FDSet *fds, const char *key, int fd) {
         return serialize_item_format(f, key, "%i", copy);
 }
 
+int serialize_fd_many(FILE *f, FDSet *fds, const char *key, const int fd_array[], size_t n_fd_array) {
+        _cleanup_free_ char *t = NULL;
+
+        assert(f);
+        assert(fd_array || n_fd_array == 0);
+
+        for (size_t i = 0; i < n_fd_array; i++) {
+                int copy;
+
+                copy = fdset_put_dup(fds, fd_array[i]);
+                if (copy < 0)
+                        return log_error_errno(copy, "Failed to add file descriptor to serialization set: %m");
```

Comment:
why error level here, when the deserialize below doesn't log?

Reply:
the log level stuff is a bit of a mess in the serialization/deserialization logic. 
 
but my original thinking was that: serialization is something we could theoretically handle, and return from, i.e. simply return to the status quo ante, and flush the partial serialization out. But deserialization is not like that: we are fucked if we can't deserialize, because we already flushed everything out, we cannot come back from that.

And then there's the upgrade thing: serialization must be clean, no excuses. It's our codebase, our logic, our consistent code. But deserialization is not like that: the serialization we get might come from older versions of systemd, and we must be permissive, lenient when parsing it.

Hence the general logic should probably be:

1. when serializing, log loudly about errors, fail if something doesn't work
2. when deserializing, only debug log about errors, try to continue as hard as possible.

Reply:
ok